### DISCUSSION TITLE
Thoughts on using Mixpanel to track potential performance issues?

### DISCUSSION THREAD

**Discussion by hirehamir** [mixpanel/mixpanel-js#452] [General]
What are your thoughts on using Mixpanel to track potential performance issues?

**Comment by hirehamir** [comment]
I'm thinking of something like this, for example.

```ts
      <pre className={styles.diffContent}>
        {fileToDiff[selectedFile].map((line, index) => {
          const lineLength = line.length;
          const startMs = performance.now();
          const renderedLine = renderLine(line, index, keywordConfigs);
          const endMs = performance.now();
          const milliseconds = endMs - startMs;
          if (milliseconds > 2) {
            mixpanel.track('Slow Line Render', {
              lineLength,
              milliseconds,
            });

            if (process.env.NODE_ENV === 'development') {
              console.debug(
                `Line with length ${lineLength} took ${milliseconds} ms to render.`
              );
            }
          }

          return renderedLine;
        })}
      </pre>
```

**Reply by tdumitrescu** [reply]
Sure, we do tracking like this internally at Mixpanel. Note that Mixpanel isn't intended as a replacement for tools that focus on operational monitoring (APM, distributed tracing, error monitoring, log analysis), and we do use dedicated tools for those purposes alongside our MP tracking. Where Mixpanel provides extra benefits alongside those tools is in cases such as: 1) you want to attach a bunch of arbitrary contextual data to your events and don't want to worry about cardinality limitations of metrics aggregation tools; 2) you want to join these monitoring events easily to other product data, for instance easily surfacing whether slow rendering correlates to other characteristics such as customer account size, data volumes, age of account etc; 3) you want to perform behavioral analysis on monitoring events that usually isn't supported by operational monitoring tools.

**Reply by hirehamir** [reply]
That makes sense. Thank you for the detailed response.
